Hello, Clojurescript Track. Again!

LeaveNhA opened the pull request:
@iHiD, I think I'm understanding now V2 (and before) and prepare the exercises. Also, my main goal is V3.

Would you check it out and tell me if I'm on the right way ?

LeaveNhA committed:
Hello World exercise added. Readme is almost done. Added myself into the
maintainers file.

iHiD commented:
This looks good now :) 

Do you want to work out how to set up Continuous Integration for this repo? At the moment it looks like it's now working.

LeaveNhA committed:
Two-fer exercise implemented.

LeaveNhA committed:
Reverse-string exercise implemented...

LeaveNhA committed:
Bob exercise implemented...

LeaveNhA committed:
Armstrong numbers exercise implemented...

LeaveNhA committed:
RNA Transcription exercise implemented...

LeaveNhA committed:
Exercise file is fixed.
Example source file added.

LeaveNhA committed:
Run Length Encoding exercise implemented...

LeaveNhA committed:
Fixing the forgotten resolved exercise code from exercise source.

LeaveNhA committed:
Yet another fix due to my notes.

LeaveNhA committed:
ISBN Verifier exercise implemented...

LeaveNhA committed:
Word Count exercise implemented...

LeaveNhA commented:
All exercises (nearly, all of the latests) comes from Clojure Track. I'll implement new test cases and rewrite the inherited code.

LeaveNhA committed:
Anagram exercise implemented...

LeaveNhA committed:
Beer Song exercise implemented...

LeaveNhA committed:
Beer Song fixed.

LeaveNhA committed:
Series exercise implemented...

LeaveNhA committed:
Nucleotide Count exercise implemented...

LeaveNhA committed:
Hamming exercise implemented...

LeaveNhA committed:
Phone Number exercise implemented...

LeaveNhA committed:
Clock exercise implemented...

LeaveNhA committed:
Leap exercise implemented...

LeaveNhA committed:
Triangle exercise implemented...

LeaveNhA committed:
Spiral Matrix exercise implemented...

LeaveNhA committed:
Raindrops exercise implemented...

LeaveNhA committed:
Track exercises added to the config file of the Track.

LeaveNhA committed:
Travis YAML file config on clojurescript track tests base.

LeaveNhA committed:
Added for track tests with Travis.

LeaveNhA committed:
Fix on travis config for language property.

LeaveNhA committed:
Yet another configuration!

LeaveNhA commented:
I'm almost bootstrap the Track.

@iHiD, I have questions;

- What is the purpose of the UUID field in every exercise data in the Track config file ?
- `configlet` does not work as we expected. I examine the shell code, I think it is OK. The Travis logs tell us about problem. I'll debug and place a solution, when I'm back home. If it's a known issue, can you throw me a solution ?
- twenty exercises are ready. Their test cases mostly comes from Clojure Track. I'll implement new cases and change the old ones.  I'll be finished it by the end of the week. Is there any problem the way it is ?

iHiD commented:
> What is the purpose of the UUID field in every exercise data in the Track config file ?

This is what allows Exercism to uniquely identify an exercise. It means that if it's other properties change (such as its name), we know that we're refering to the same exercise. Once merged, a UUID can never change.

> configlet does not work as we expected. I examine the shell code, I think it is OK. The Travis logs tell us about problem. I'll debug and place a solution, when I'm back home. If it's a known issue, can you throw me a solution ?

I'm not sure about this, but I've added #4 which maybe fixes it? If not, I'll ask on Slack for you and someone might know.

> twenty exercises are ready. Their test cases mostly comes from Clojure Track. I'll implement new cases and change the old ones. I'll be finished it by the end of the week. Is there any problem the way it is ?

This sounds amazing 💙

iHiD commented:
I've added @ErikSchierboom to give this a check once he's back next week 🙁

LeaveNhA committed:
Re-factored, Hello World exercise test.

LeaveNhA committed:
Added example on Hello World exercise.

LeaveNhA committed:
Hello World Exercise UUID added in config file.

LeaveNhA committed:
Added Test and Submit fields in README document about exercise.

LeaveNhA committed:
Typo fixed.

LeaveNhA committed:
Re-factored Two-Fer exercise tests...

LeaveNhA committed:
Added Two-Fer exercise UUID to config file.

LeaveNhA committed:
Re-factored Reverse-String example source with restrictions of the exercise.

LeaveNhA committed:
Typo and explanation fix on Reverse-String exercise file.

LeaveNhA committed:
Running Tests and Submitting fields added to the Reverse-String exercise.

LeaveNhA committed:
Added one external test. Re-factored some of the old ones.
All tests are rewritten.

LeaveNhA committed:
UUID for Reverse-String exercise added to the config file.

LeaveNhA committed:
Nearly, nothing has changed, tiny-little new-line added for source.

LeaveNhA committed:
Fix on Bob source. Removed forgotten exercise bootstrap-test sources.

LeaveNhA committed:
Re-factored test cases of Bob the exercise.

LeaveNhA committed:
UUID added for exercise Bob.

LeaveNhA committed:
Armstrong Numbers readme file added new-line for source.

LeaveNhA committed:
Fixed typo.

LeaveNhA committed:
Armstrong Numbers exercise tests are re-factored.

LeaveNhA committed:
UUID added for Armstrong Numbers exercise.

LeaveNhA committed:
Typo fixed on RNA Transcription exercise source.

LeaveNhA committed:
Re-factored RNA Transcription exercise example source.

LeaveNhA committed:
Re-factored RNA Transcription exercise tests.

LeaveNhA committed:
Added UUID for RNA Transcription exercise.

LeaveNhA committed:
Added UUID for Run-Length-Encoding exercise.

LeaveNhA committed:
Typo fixed.

LeaveNhA committed:
Changed the exercise example solution.

LeaveNhA committed:
Typo fixed...

LeaveNhA committed:
Re-factored Run Length Encoding exercise tests.

LeaveNhA committed:
Added UUID for ISBN Verifier exercise.

LeaveNhA committed:
Typo fixed.

LeaveNhA committed:
Typo fixed.

LeaveNhA committed:
Re-factored the example of ISBN Verifier exercise.

LeaveNhA committed:
Re-factored ISBN Verifier tests.

LeaveNhA committed:
Added UUID for Word Count exercise.

LeaveNhA committed:
New-line added for source.

LeaveNhA committed:
Re-factored Word Count exercise example source.

LeaveNhA committed:
Typo fixed.

LeaveNhA committed:
Re-factored Word Count exercise test.

LeaveNhA committed:
New-line added to Anagram exercise readme source.

LeaveNhA committed:
Added UUID for Anagram exercise.

LeaveNhA commented:
> I've added @ErikSchierboom to give this a check once he's back next week 🙁

There is no rush on me, I'll stick my plan. I want all of the exercises. 😇👨🏻‍💻

LeaveNhA committed:
Re-factored Beer Song exercise completely;
Added UUID for the exercise to the exercises config file.
Re-factored example source.
re-factored test source.

LeaveNhA committed:
Removed the forgotten dead-codes from test source.

bobbicodes reviewed (approved):
Wow, this looks amazing!

LeaveNhA commented:
> Wow, this looks amazing! 

Thank you! 😇🙋🏻‍♂️

LeaveNhA committed:
Series exercise re-factored...
Added exercise UUID.
Re-factored example source.
Re-factored exercise source.
Re-factored test.

LeaveNhA committed:
Re-factored Nucleotide Count exercise...
Added UUID into config file.
Re-factored example source.
Re-factored exercise source.
Re-factored and enhanced the test of the exercise.

LeaveNhA committed:
Re-factored Hamming exercise...
Added UUID to the config file.
Re-factored example source.
Re-factored exercise source.
Re-factored and enhanced the test source.

bobbicodes commented:
The [Babashka Track](https://github.com/exercism/babashka/) may be of reference, since I had to get rid of anything using JVM-specific features and doing so resulted in more portable code. So porting from Babashka to Clojurescript might even be easier.

LeaveNhA committed:
Re-factored Clock exercise...
Added UUID to the config file.
Re-factored example source.
Re-factored tests. (I'll enhance them in the next week, TODO!)

LeaveNhA committed:
Re-factored Leap exercise...
Added UUID for the exercise.
Fixed typo on source with README.md.
Re-factored the example source.
Re-factored the test source.

LeaveNhA committed:
Removed duplicated exercise object.

LeaveNhA committed:
Re-factored Raindrops exercise...
Added UUID for the exercise.
Re-factored example source.
Re-factored exercise source.
Re-factored test source.

LeaveNhA commented:
> The [Babashka Track](https://github.com/exercism/babashka/) may be of reference, since I had to get rid of anything using JVM-specific features and doing so resulted in more portable code. So porting from Babashka to Clojurescript might even be easier.

Thank you. I'll consider it. 😇

LeaveNhA commented:
Last two exercise to re-factor and then we are ready to Launch! 😍😇🙏🏻🥳👨🏻‍💻

LeaveNhA committed:
Re-factored Spiral Matrix exercise...
Fixed a newline on README.md file.
Re-factored example source.
Re-factored exercise source.
Re-factored exercise test.

LeaveNhA committed:
Re-factored Triangle exercise.
Added UUID for the exercise to the config file.
Re-factored exercise test.

ErikSchierboom commented on .travis.yml:
```diff
@@ -1,5 +1,15 @@
 script:
-  - bin/fetch-configlet
-  - bin/configlet lint --track-id clojurescript .
+- bin/fetch-configlet
```
The `fetch-configlet` version that is in this repository is an older one that has an issue, namely that it doesn't use authentication when retrieving the binary from GitHub. This leads to GitHub rate limits being hit from time to time, causing otherwise perfectly valid builds to fail (see [this issue for a description](https://github.com/exercism/configlet/issues/173)). 

The issue was solved in [this PR](https://github.com/exercism/configlet/pull/174) and it updates the `fetch-configlet` binary to use authentication to retrieve the binaries. The only disadvantage of this method is that it requires the GitHub credentials to be stored somewhere on the build server as a secret. For Travis, this is a bit of manual work, but for GitHub actions, this is done seamlessly.

The C# track use GitHub actions as its CI and the updated fetch-configlet script: https://github.com/exercism/csharp/blob/master/.github/workflows/test.yml Would you be interested in converting the build to use GitHub actions? For the C# track, I've found that switching to GH actions also lead to a substantial performance improvement. 

I'd be happy to help if you think this is something you want to do.

LeaveNhA replied:
Secret is wanted. But the code-part is ready.

ErikSchierboom commented on config.json:
```diff
@@ -1,10 +1,176 @@
+    "language": "ClojureScript",
+    "track_id": "clojurescript",
+    "active": true,
+    "blurb": "",
```
Could you perhaps add a blurb? It is a short description of the language. See e.g. https://github.com/exercism/csharp/blob/master/config.json#L4

LeaveNhA replied:
Added. Thanks. 😇

ErikSchierboom reviewed (commented):
This is great! I've left two smaller comments.

One thing you should do before launching is to add contents to the documentation files in the [docs folder](https://github.com/exercism/clojurescript/tree/master/docs), as students rely on those documents to get started with the track (setting up their system etc.).

ErikSchierboom commented on config.json:
```diff
@@ -1,10 +1,176 @@
+            "difficulty": 1,
+            "unlocked_by": null,
+            "auto_approve": true,
+            "core": false,
```
@iHiD I don't know exactly what happens if all the exercises have `core` set to `false`. Do you? We probably don't want to make too much of an effort to order the exercise with unlocks and such, right? Or maybe a small pass.

LeaveNhA replied:
I've solved it, but we should discuss it in the new PR.
Thanks again. 😇
